src/pyodide-py/pyodide/webloop.py

```diff
@@ -0,0 +1,425 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+# flake8: noqa
+
+"""
+
+A custom event loop for running asyncio in Pyodide
+
+Version: 0.1.2
+
+Author: Wei OUYANG (@oeway)
+        Adapted from the EventSimulator made by @damonjw
+        https://gist.github.com/damonjw/35aac361ca5d313ee9bf79e00261f4ea
+
+Version History:
+        * support contextvars
+        * add event loop policy
+        * throw error when the loop is already running
+        * support Python 3.8
+        * fix run_until_complete
+"""
+
+import heapq
+import asyncio
+import time
+import traceback
+import contextvars
+
+import js
+from typing import Dict, Tuple, Optional, Awaitable, Callable
+
+
+class WebLoop(asyncio.AbstractEventLoop):
+    """A custom event loop for running asyncio in Pyodide
+
+    It works by utilizing the browser event loop via the setTimeout function
+    """
+
+    def __init__(self, debug: Optional[bool] = False, interval: Optional[int] = 10):
+        """
+        Instatiate the web loop
+        """
+        self._running = False
+        self._immediate = []
+        self._scheduled = []
+        self._futures = []
+        self._next_handle = None
+        self._debug = debug
+        self._stop = False
+        self._interval = interval
+        self._timeout_promise = js.eval(
+            "self._timeoutPromise = function(time){return new Promise((resolve)=>{setTimeout(resolve, time);});}"
+        )
+        self._exception_handler = None
+        self._task_factory = self._default_task_factory
+        self._stop_callbacks = None
+        self._result = None
+        self._exception = None
+
+    def get_debug(self):
+        """
+        Get the debug mode (bool) of the event loop.
+        """
+        return self._debug
+
+    def set_debug(self, enabled: bool):
+        """
+        Set the debug mode of the event loop.
+        """
+        self._debug = enabled
+
+    def time(self):
+        """
+        Return the current time, as a float value, according to the time module (time.time()).
+        """
+        return time.time()
+
+    def run_forever(self):
+        """
+        Run the event loop until stop() is called.
+
+        Note that this function is different from the standard asyncio loop implementation in two ways:
+         1) It won't block the execution
+         2) It returns a Promise object
```

Comment:
Please explain here under what conditions the promise will return, under what conditions it will raise, what values will it return, what will it raise.